go/src/pluginmgr/pluginmgr_common/pluginmgr_helper.go

```diff
@@ -0,0 +1,740 @@
+					lomcommon.DurationToSeconds(PLUGIN_INIT_PERIODIC_TIMEOUT_DEFAULT), "")
+				isError = true
+			}
+			if err != nil {
+				plmgr.AddPeriodicLogNotice(uid, fmt.Sprintf("Failed to initialize plugin %s: %s", pluginName, err.Error()),
+					lomcommon.DurationToSeconds(PLUGIN_INIT_PERIODIC_TIMEOUT_DEFAULT), "")
+				isError = true
+			}
+			if isError {
+				initTimeoutCalled = false
+				if !initTimeout.Stop() {
+					<-initTimeout.C
+				}
+				logTimer.Reset(PLUGIN_INIT_PERIODIC_TIMEOUT_DEFAULT) // log very frequently every 5min default
+				continue
+			}
+			if time.Since(initstartTime) <= (PLUGIN_INIT_CALL_TIMEOUT_DEFAULT + 1*time.Second) {
+				// Got some non error status from init() within 60sec (default), so plujgin is sucecssfully initialized.
+				// To-Do : Goutham : Define & Change remaining metadata fields
+				pluginMetadata := plugin.GetMetadata()
+				pluginMetadata.StartedTime = time.Now()
+				plugin.SetMetadata(pluginMetadata)
+				plmgr.plugins[pluginName].SetPluginStage(plugins_common.PluginStageInitSuccess)
+
+				// Call register action with engine
+				goroutinetracker.Start("plg_mgr_AddPlugin_register"+pluginName,
+					plmgr.RegisterActionWithEngine(pluginName), pluginName)
+				return nil
+
+			} else {
+				// Got success from plugin's init() after 60 seconds(default time). So discard this status and clear any periodic logging. Treat it like Plugin initialization failed
+				pluginMetadata := plugin.GetMetadata()
+				pluginMetadata.StartedTime = time.Now()
+				plugin.SetMetadata(pluginMetadata)
+				plmgr.RemovePeriodicLogEntry(uid)
+				plmgr.plugins[pluginName].SetPluginStage(plugins_common.PluginStageInitFailure)
+				//delete(plmgr.plugins, pluginName) // Verify : Goutham : Is this needed instead?
+				return lomcommon.LogError("Failed to initialize plugin %s: init() returned success after 60sec(default)", pluginName)
+			}
+		case <-logTimer.C: // Fired after every 5min (default)
+			if logfreq == 0 || logfreq%12 == 0 {
+				plmgr.UpdatePeriod(uid, lomcommon.DurationToSeconds(PLUGIN_INIT_PERIODIC_FALLBACK_TIMEOUT_DEFAULT)) // fall back from frequent logging to 1Hr(default) window to avoid syslog pollution
+				logfreq = logfreq % 12
+				if !logTimer.Stop() {
+					<-logTimer.C
+				}
+				if !initTimeoutCalled {
+					break // only break if there is error returned from plugin's init()
+				}
+			}
+			logfreq++
+		case <-initTimeout.C: // Fired after 60sec(default) , init() didn't returned anything
+			initTimeoutCalled = true
+			plmgr.AddPeriodicLogNotice(uid, fmt.Sprintf("Timed out on waiting for plugin %s to initialize", pluginName),
+				lomcommon.DurationToSeconds(PLUGIN_INIT_PERIODIC_TIMEOUT_DEFAULT), "")
+			logTimer.Reset(PLUGIN_INIT_PERIODIC_TIMEOUT_DEFAULT) // log very frequently every 5 min(default)
+		}
+	}
+}
+
+func (plmgr *PluginManager) RegisterActionWithEngine(pluginName string) error {
```

Comment:
-  suggestion to make this unexported method since other packages are not expected to use it. If its needed for tests thats fine.

Reply:
Yes. I made all of plugin managers methods private for now. main() just calls setup methods.